### PR TITLE
checkout: fix partial file checkouts in the same directory

### LINKED ISSUE #5158: pull/checkout: pulling multiple files consecutively from a folder checks out only first
# Bug Report

## Description

When attempting to pull individual files from dvc tracked folder, only the first pull creates file links. Similar behavior with `checkout` unless `--relink` is used.

### Reproduce
0. (git + dvc init)

1. mkdir db
2. touch db/a.txt && touch db/b.txt
3. dvc add db
4. git add db.dvc && git commit -m "dvc"
5. dvc remote add -d tmp /tmp/remote
6. dvc push
7. rm db/*
8. dvc pull db/a.txt
9. dvc pull db/b.txt

### Expected

db/a.txt & db/b.txt exist (When in fact b.txt doesn't). 

Getting messages:
dvc pull db/a.txt --> 1 file modified
dvc pull db/b.txt --> Everything is up to date. 

### Environment information

DVC version: 1.11.9 (pip)
---------------------------------
Platform: Python 3.6.10 on Linux-4.18.0-1024-azure-x86_64-with-debian-buster-sid
Supports: http, https, s3
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sda1
Caches: local
Remotes: local
Workspace directory: ext4 on /dev/sda1
Repo: dvc, git

### PR DESCRIPTION
**Pull request by isidentical**
Resolves #5158

### CODE REVIEW

File: tests/func/test_checkout.py [hunk 1]
```diff
@@ -68,6 +68,37 @@ def test(self):
...
+
+        file1, file2 = files
+
+        self.dvc.checkout(targets=file1)
+        self.assertTrue(os.path.exists(self.CO_DIR))
+        self.assertTrue(os.path.exists(file1))
+        self.assertFalse(os.path.exists(file2))
+
+        self.dvc.checkout(targets=file2)
+        self.assertTrue(os.path.exists(file2))
```
Comment (efiop) [inline tests/func/test_checkout.py:99]: We are currently continuing to migrate from unittest-style tests to pytest-style tests https://github.com/iterative/dvc/issues/1819 , so all new tests need to be in pytest-style. Could you please rewrite it? E.g. see examples at the bottom of this test_checkout.py file.

Comment (efiop) [reply tests/func/test_checkout.py:99]: Also, we need to test a case for checking out a subdirectory in a tracked directory, that might break.

File: dvc/cache/base.py [hunk 2]
```diff
@@ -520,7 +520,9 @@ def checkout(
             self.safe_remove(path_info, force=force)
             failed = path_info
 
-        elif not relink and not self.changed(path_info, hash_info):
+        elif not relink and not self.changed(
+            filter_info or path_info, hash_info
+        ):
```
Comment (efiop) [inline dvc/cache/base.py:525]: hash_info is a hash_info of the full dataset, right? But filter_info might be about a subset. Please check `changed()` method, I don't think it is ready for this.

File: tests/func/test_checkout.py (hunk 1 above)
Comment (isidentical) [reply tests/func/test_checkout.py]: Applied both of them

File: dvc/cache/base.py (hunk 2 above)
Comment (isidentical) [reply dvc/cache/base.py]: You are right. In 323d20d, I tried to address by using filter_info in changed if the checks related to paths, and using the original path_info otherwise (like comparing the hashs)

File: dvc/cache/base.py [hunk 3]
```diff
@@ -101,26 +101,28 @@ def changed(self, path_info, hash_info):
         Args:
             path_info: dict with path information.
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific file.
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        real_path_info = filter_info or path_info
```
Comment (skshetry) [inline dvc/cache/base.py:110]: `real` *adj* might be confusing here. How about just `path`?

File: tests/func/test_checkout.py [hunk 4]
```diff
@@ -880,3 +880,31 @@ def test_checkout_executable(tmp_dir, dvc):
...
+        db_dir / filename for filename in ("a.txt", "b.txt", sub_dir / "c.txt")
+    ]
+
+    shutil.rmtree(db_dir)
+
+    dvc.checkout(targets=[str(file_1)])
+    assert db_dir.exists()
+    assert file_1.exists()
+    assert not file_2.exists()
+    assert not sub_dir.exists()
```
Comment (skshetry) [inline tests/func/test_checkout.py:902]: There's an easier way to express this in a more succinct way:
```suggestion
    assert db_dir.read_text() == {"a.txt": "a"}
```
This will check the contents as well as the folder.

File: tests/func/test_checkout.py [hunk 5]
```diff
@@ -880,3 +880,31 @@ def test_checkout_executable(tmp_dir, dvc):
...
+
+    dvc.checkout(targets=[str(sub_dir_file)])
+    assert sub_dir.exists()
+    assert sub_dir_file.exists()
+    assert not file_2.exists()
```
Comment (skshetry) [inline tests/func/test_checkout.py:907]: Similarly as https://github.com/iterative/dvc/pull/5163/files#r549324917,
```suggestion
    assert db_dir.read_text() == {"a.txt": "a", "sub_dir": {"c.txt": "c"}}
```

File: tests/func/test_checkout.py [hunk 6]
```diff
@@ -880,3 +880,31 @@ def test_checkout_executable(tmp_dir, dvc):
...
+
+
+def test_checkout_partial(tmp_dir, dvc):
+    tmp_dir.dvc_gen(
+        {"db": {"a.txt": "a", "b.txt": "b", "sub_dir": {"c.txt": "c"}}}
+    )
+
+    db_dir = tmp_dir / "db"
+    sub_dir = db_dir / "sub_dir"
+    file_1, file_2, sub_dir_file = [
```
Comment (skshetry) [inline tests/func/test_checkout.py:892]: Should we also add a test for `sub_dir`?

File: tests/func/test_checkout.py [hunk 7]
```diff
@@ -880,3 +880,31 @@ def test_checkout_executable(tmp_dir, dvc):
...
+        db_dir / filename for filename in ("a.txt", "b.txt", sub_dir / "c.txt")
+    ]
+
+    shutil.rmtree(db_dir)
+
+    dvc.checkout(targets=[str(file_1)])
```
Comment (skshetry) [inline tests/func/test_checkout.py:898]: *Nitpick*: `checkout` supports a single `str` argument as well. 
```suggestion
    dvc.checkout(targets=str(file_1))
```

File: tests/func/test_checkout.py (hunk 5 above)
Comment (isidentical) [reply tests/func/test_checkout.py]: I wasn't aware of this functionality, thanks @skshetry for letting me know!

File: tests/func/test_checkout.py (hunk 6 above)
Comment (isidentical) [reply tests/func/test_checkout.py:892]: Can you be more explicit about the case? We already have a test case to checkout a partial file from a subdirectory but if you mean anything, we can add that too!

File: tests/func/test_checkout.py (hunk 7 above)
Comment (isidentical) [reply tests/func/test_checkout.py]: ah, cool, thanks!

File: dvc/cache/base.py [hunk 8]
```diff
@@ -101,24 +101,26 @@ def changed(self, path_info, hash_info):
         Args:
             path_info: dict with path information.
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific file.
```
Comment (efiop) [inline dvc/cache/base.py:104]: filter_info could also target a subdir, not only a sepcific file

File: dvc/cache/base.py [hunk 9]
```diff
@@ -101,24 +101,26 @@ def changed(self, path_info, hash_info):
...
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        path = filter_info or path_info
         logger.trace(
-            "checking if '%s'('%s') has changed.", path_info, hash_info
+            "checking if '%s' ('%s') has changed.", path_info, hash_info
```
Comment (efiop) [inline dvc/cache/base.py:112]: Minor: this is an unrelated change. This should either be a separate commit or should not be in this PR at all. 
```suggestion
            "checking if '%s'('%s') has changed.", path_info, hash_info
```

File: dvc/cache/base.py [hunk 10]
```diff
@@ -101,24 +101,26 @@ def changed(self, path_info, hash_info):
...
 
         if not hash_info:
             logger.debug("hash value for '%s' is missing.", path_info)
             return True
 
-        if self.changed_cache(hash_info):
+        if self.changed_cache(hash_info, filter_info=filter_info):
             logger.debug(
                 "cache for '%s'('%s') has changed.", path_info, hash_info
             )
```
Comment (efiop) [inline dvc/cache/base.py:126]: In the logic below we compare `get_hash(path_info)` and `hash_info`, which means that if the directory you are trying to checkout a subdir/subfile of has some unrelated subdir/subfile changed, this method will think that the subdir/subfile is changed and will checkout it again. That's not right.

File: tests/func/test_checkout.py (hunk 6 above)
Comment (efiop) [reply tests/func/test_checkout.py:892]: @isidentical I believe @skshetry is saying that we should test `dvc checkout db/sub_dir` too. I left a comment about it as well.

File: dvc/cache/base.py (hunk 10 above)
Comment (isidentical) [reply dvc/cache/base.py:126]: From what I understand, we only have access to the hash_info that already presented in the dvcfile which is usually the base directory. So, this behavior seems correct yet inefficient but I failed to find any better alternative with the current dvcfile format. Do you have any hints for me regarding this?

File: tests/func/test_checkout.py (hunk 6 above)
Comment (isidentical) [reply tests/func/test_checkout.py:892]: Oh, right! Sure, can add it.

File: dvc/cache/base.py (hunk 10 above)
Comment (efiop) [reply dvc/cache/base.py:126]: @isidentical Take a look at `set_dir_info` helper. It tries to load `hash_info.dir_info` if the full info for `hash_info` is present in the cache. We could use that too here (or maybe `get_dir_cache` directly) to get a more complete info, then filter it out and use it for a correct comparison. If dir cache is not available - we could indeed stick to the current behavior as a fallback.

File: dvc/cache/base.py [hunk 11]
```diff
@@ -101,40 +101,52 @@ def changed(self, path_info, hash_info):
...
             bool: True if data has changed, False otherwise.
         """
 
+        hash_origin = path_info
+        if filter_info:
+            path = filter_info
+            dir_info = self.get_dir_cache(hash_info)
+            hash_key = path.relative_to(path_info).parts
+            if hash_key in dir_info.trie:
+                hash_info = dir_info.trie[hash_key]
```
Comment (efiop) [inline dvc/cache/base.py:116]: Are you sure this works correctly for a subdirectory?

Comment (efiop) [reply dvc/cache/base.py:116]: Looks like this will be skipped for a subdir, right? But you can actually compute the hash for the subdir if you have hashes for all the files inside of it. Take a look at how we compute hash_info for the directory.

Comment (isidentical) [reply dvc/cache/base.py]: I tried to address this problem in the 1ced36c, would you mind checking the latest revision out?

File: dvc/cache/base.py [hunk 12]
```diff
@@ -101,40 +117,53 @@ def changed(self, path_info, hash_info):
...
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        hash_origin = path_info
+        path = filter_info or path_info
+
         logger.trace(
             "checking if '%s'('%s') has changed.", path_info, hash_info
```
Comment (efiop) [inline dvc/cache/base.py:130]: Needs update

File: dvc/cache/base.py [hunk 13]
```diff
@@ -101,40 +117,53 @@ def changed(self, path_info, hash_info):
...
             return True
 
         if not hash_info:
             logger.debug("hash value for '%s' is missing.", path_info)
             return True
 
-        if self.changed_cache(hash_info):
+        if self.changed_cache(hash_info, filter_info=filter_info):
             logger.debug(
                 "cache for '%s'('%s') has changed.", path_info, hash_info
```
Comment (efiop) [inline dvc/cache/base.py:143]: will need update too

File: dvc/cache/base.py [hunk 14]
```diff
@@ -101,40 +117,53 @@ def changed(self, path_info, hash_info):
...
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific path.
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        hash_origin = path_info
+        path = filter_info or path_info
+
```
Comment (efiop) [inline dvc/cache/base.py:128]: The naming is really confusing. Same with `_get_specialized_hash_info`

File: dvc/cache/base.py [hunk 15]
```diff
@@ -101,40 +117,53 @@ def changed(self, path_info, hash_info):
...
+        # Instead of directly checking whether filter_info is present or not,
+        # we compare the target path with the path_info because sometimes
+        # filter_info might be equal to the path_info and in those cases we do
+        # not need to specialize the hash.
+        if path != path_info:
+            hash_origin, hash_info = self._get_specialized_hash_info(
+                path, hash_info, hash_origin
+            )
+
+        actual = self.tree.get_hash(hash_origin)
```
Comment (efiop) [inline dvc/cache/base.py:156]: The logic here is really confusing. Maybe just filter the hash and then use a helper method to check if the target has changed.

File: dvc/cache/base.py [hunk 16]
```diff
@@ -89,7 +89,23 @@ def load_dir_cache(self, hash_info):
...
+        else:
+            depth = len(hash_key)
+            specialized_dir_info = DirInfo()
+            try:
+                for key, value in dir_info.trie.items(hash_key):
+                    specialized_dir_info.trie[key[depth:]] = value
+            except KeyError:
+                return hash_origin, hash_info
+            else:
+                return path_info, self._get_dir_info_hash(specialized_dir_info)
```
Comment (efiop) [inline dvc/cache/base.py:106]: This method is really confusing and is doing too much. The fallback logic here doesn't really apply to the filtered hash_info (btw `filtered` is probably a better term than `specialized` here).

Also note how giant the `else` statement is after `if & return`. We don't have this check enabled in pylint yet, but we do appreciate no-else-after-return unless there is a good reason to do otherwise.

**Review by efiop — COMMENTED** [review]
We are on the right path, but need to clean up the code a bit :slightly_smiling_face:

File: dvc/cache/base.py (hunk 15 above)
Comment (isidentical) [reply dvc/cache/base.py]: I've kept the comparison logic in place, I think it seems a bit more simple now since we moved to the filtering part to the upper side. Could you check it again? (if you still find it complicated, I can move it out as well, as a separate function)

File: dvc/cache/base.py [hunk 17]
```diff
@@ -89,7 +89,29 @@ def load_dir_cache(self, hash_info):
 
         return DirInfo.from_list(d)
 
-    def changed(self, path_info, hash_info):
+    def _get_filtered_hash_info(
+        self, filtered_path_info, hash_info, hash_origin
```
Comment (efiop) [inline dvc/cache/base.py:93]: Doesn't seem like we need to introduce alternative names, only makes it more confusing
```suggestion
        self, filter_info, hash_info, path_info
```

File: dvc/cache/base.py [hunk 18]
```diff
@@ -101,40 +123,53 @@ def changed(self, path_info, hash_info):
         Args:
             path_info: dict with path information.
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific path.
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        hash_origin = path_info
```
Comment (efiop) [inline dvc/cache/base.py:132]: This assignment is probably not needed. You can probably just use `path_info`.

Comment (isidentical) [reply dvc/cache/base.py:132]: Not really, this is the origin of the `hast_info`. If we can successfully filter it, we change it below to the `filter_info`. And when comparing hashes, we use this location for obtaining the current cache

File: dvc/cache/base.py [hunk 19]
```diff
@@ -101,17 +121,26 @@ def changed(self, path_info, hash_info):
...
             bool: True if data has changed, False otherwise.
         """
 
+        if filter_info is not None and filter_info != path_info:
+            path = filter_info
+            path_info, hash_info = self._get_filtered_hash_info(
+                path, hash_info, path_info
+            )
+        else:
+            path = path_info
```
Comment (efiop) [inline dvc/cache/base.py:136]: Notice how we don't really need filtered hash_info until `changed_cache` call below. We shouldn't try to filter it until we really need it, as it might be costly for big datasets. The `trace()` message that uses it can be removed or modified to just include filter_info too, so you don't have to have filtered hash_info for it.

File: dvc/cache/base.py [hunk 20]
```diff
@@ -89,7 +89,27 @@ def load_dir_cache(self, hash_info):
...
+        filtered_dir_info = DirInfo()
+        try:
+            for key, value in dir_info.trie.items(hash_key):
+                filtered_dir_info.trie[key[depth:]] = value
+        except KeyError:
+            # The directory doesn't exist on the trie, fallback to the original
+            return path_info, hash_info
+        else:
+            filtered_hash_info, _ = self._get_dir_info_hash(filtered_dir_info)
+            return filter_info, filtered_hash_info
```
Comment (efiop) [inline dvc/cache/base.py:110]: Note how the first return value is either `filter_info` or `path_info`(just a binary switch). There should be a nicer way to handle this.

Comment (isidentical) [reply dvc/cache/base.py]: They both are `PathInfo`s, and I don't know what else I can return there. Maybe just raise an exception when we can't filter (just like a signal) and then use catch it right away?

File: dvc/cache/base.py [hunk 21]
```diff
@@ -89,7 +89,26 @@ def load_dir_cache(self, hash_info):
...
+        depth = len(hash_key)
+        filtered_dir_info = DirInfo()
+        try:
+            for key, value in dir_info.trie.items(hash_key):
+                filtered_dir_info.trie[key[depth:]] = value
+        except KeyError:
+            return None
+        else:
+            filtered_hash_info, _ = self._get_dir_info_hash(filtered_dir_info)
+            return filtered_hash_info
```
Comment (efiop) [inline dvc/cache/base.py:109]: Just a bit flatter, since we `return` in `except` anyway.
```suggestion

        filtered_hash_info, _ = self._get_dir_info_hash(filtered_dir_info)
        return filtered_hash_info
```

File: dvc/cache/base.py [hunk 22]
```diff
@@ -89,7 +89,26 @@ def load_dir_cache(self, hash_info):
...
+        depth = len(hash_key)
+        filtered_dir_info = DirInfo()
+        try:
+            for key, value in dir_info.trie.items(hash_key):
+                filtered_dir_info.trie[key[depth:]] = value
+        except KeyError:
+            return None
+
+        filtered_hash_info, _ = self._get_dir_info_hash(filtered_dir_info)
+        return filtered_hash_info
```
Comment (efiop) [inline dvc/cache/base.py:109]:
```suggestion
        return self._get_dir_info_hash(filtered_dir_info)[0]
```

File: dvc/cache/base.py [hunk 23]
```diff
@@ -101,23 +120,33 @@ def changed(self, path_info, hash_info):
...
-            logger.debug("'%s' doesn't exist.", path_info)
+        if not self.tree.exists(path):
+            logger.debug("'%s' doesn't exist.", path)
             return True
 
         if not hash_info:
             logger.debug("hash value for '%s' is missing.", path_info)
             return True
 
+        if path != path_info:
```
Comment (efiop) [inline dvc/cache/base.py:142]:
```suggestion
        if filter_info:
```

Comment (isidentical) [reply dvc/cache/base.py:142]: This is not always true, since filter_info might equal to the path_info sometimes.

Comment (efiop) [reply dvc/cache/base.py:142]: ah, good point!

File: dvc/cache/base.py [hunk 24]
```diff
@@ -101,23 +120,33 @@ def changed(self, path_info, hash_info):
...
+            filtered_hash_info = self._get_filtered_hash_info(
+                path, hash_info, path_info
+            )
+            if filtered_hash_info is not None:
+                path_info = filter_info
```
Comment (efiop) [inline dvc/cache/base.py:147]: Might be missing something, but how about we don't do this assignment and just use `path` down below to avoid confusion? 

We'll need to do something like `if filtered_hash_info is None: path = path_info` as a fallback though.

Comment (efiop) [reply dvc/cache/base.py:147]: @isidentical `path_info` is coming in as an argument, so feels a bit weird to have `path` and then also overwrite `path_info`, which results in a confusion of not having to figure out if `path_info` is what came to us or is the new filtered value.

Comment (isidentical) [reply dvc/cache/base.py:147]:
```
        if path != path_info:
            filtered_hash_info = self._get_filtered_hash_info(
                path, hash_info, path_info
            )
            if filtered_hash_info is None:
                path = path_info
            else:
                hash_info = filtered_hash_info
```
something like this?

**Review by efiop — APPROVED** [review]
Thank you!
